### ISSUE TITLE
[week 2] Parallel training과 GPU를 학습에 사용하는 것은 같은 방식일까

### ISSUE THREAD

**Issue by HanNayeoniee** [kimcando/BoostcampAITech3-PaperReading-Embedding#11]
본 논문 **2.3 Parallel training of Neural Networks** 문단에서 대용량 데이터셋을 모델에 학습시키기 위해 대용량 분산 프레임워크 DistBelief를 사용했다고 하는데요, 여기서 모델을 parallel하게 학습시키기 위한 **parallel training**이 GPU를 학습에 사용하는 방식과 동일한 학습 방식인가요?

gradient update를 했을 때 모든 gradient가 update된다는 점에서는 동일한 것 같은데, parallel하게 학습시키기 위해 동일한 모델의 복제본(replica)을 만드는 부분은 다른 것 같아서 질문드립니다.
 
> The framework(DistBelief) allows us to run multiple replicas of the same model in parallel, and each replica synchronizes its gradient updates through a centralized server that keeps all the parameters. 
For this parallel training, we use mini-batch asynchronous gradient descent with an adaptive learning rate procedure called Adagrad [7].

**Comment by kimcando** [comment]
나연님 질문이 제가 이해한바가 맞는지 모르겠는데요!
multi gpu(또는 저 시대였다면 multi cpu로?) 사용한 parallel training을 말하는 것 같습니다. 보통 병렬화에서 모델 병렬화와 데이터 병렬화가 있는데요. 
* 데이터 병렬화경우 같은 동일한 모델을 각 노드(gpu) 마다 보내주고 학습 데이터를 병렬화에 사용하는 노드 갯수만큼 쪼개서 노드에 보내줍니다. 
* 그리고 각각 계산한 그래디언트를 모아서 모델을 업데이트하고, 그 업데이트된 모델을 다시 각 디바이스에 보내주고요.( 이때 각 디바이스의 그래디언트를 모으고 업데이트를 수행하는 노드를 coordinating node라고 합니당)
* 파이토치 distributeddataparallel 나 dataparallel가 요런 과정이라서 파이토치 공식페이지 참고하셔도 좋을 것 같고.. [당근마켓에 이 분이 올리신 글](https://medium.com/daangn/pytorch-multi-gpu-%ED%95%99%EC%8A%B5-%EC%A0%9C%EB%8C%80%EB%A1%9C-%ED%95%98%EA%B8%B0-27270617936b) 도 살짝 봐도 좋을 것 같아요!

반면 모델 병렬화는 큰 모델을 부분마다 쪼개서 디바이스에 보내주는걸로 알고 있어요. 한 디바이스에 한 모델을 다 담기 어려운 메모리 이슈나 모델 구간마다 계산되는 과정을 병렬화하기 위함이랄까요??( OS 수업에 Pipeline stall 처럼..) [여기](https://runebook.dev/ko/docs/pytorch/pipeline) 그림 참고해보세용!
